File: mneme-commercial-m0/services/console_api/src/mneme_console_api/audit.py
"""Audit feed endpoint."""

from __future__ import annotations

import json
from typing import Any

from .db import get_pool


async def list_events(
    event_type: str | None = None,
    subject: str | None = None,
    limit: int = 200,
) -> dict[str, Any]:
    pool = await get_pool()
    where: list[str] = []
    params: list[Any] = []
    if event_type:
        params.append(event_type)
        where.append(f"event_type = ${len(params)}")
    if subject:
        params.append(subject)
        where.append(f"subject = ${len(params)}")
    where_clause = ("WHERE " + " AND ".join(where)) if where else ""
    params.append(limit)
    sql = f"""
        SELECT id, event_type, subject, target, metadata, occurred_at
        FROM audit_events
        {where_clause}
        ORDER BY occurred_at DESC
        LIMIT ${len(params)}
    """
    async with pool.acquire() as conn:
        rows = await conn.fetch(sql, *params)
    return {
        "count": len(rows),
        "events": [
            {
                "id": r["id"],
                "event_type": r["event_type"],
                "subject": r["subject"],
                "target": r["target"],
                "metadata": _loads(r["metadata"]),
                "occurred_at": r["occurred_at"].isoformat(),
            }
            for r in rows
        ],
    }


def _loads(v):
    if v is None:
        return None
    if isinstance(v, str):
        return json.loads(v)
    return v

File: mneme-commercial-m0/apps/console/src/api/audit.ts
import { api } from './client';

export type AuditEvent = {
  id: number;
  event_type: string;
  subject: string;
  target: string | null;
  metadata: any;
  occurred_at: string;
};

export function getAudit(filters: { event_type?: string; subject?: string }, user: string) {
  const qs = new URLSearchParams();
  if (filters.event_type) qs.set('event_type', filters.event_type);
  if (filters.subject) qs.set('subject', filters.subject);
  const suffix = qs.toString() ? `?${qs}` : '';
  return api<{ count: number; events: AuditEvent[] }>(`/api/audit${suffix}`, user);
}
